Improve deployment workload support in CRUD benchmarking framework

diamondpowell opened the pull request:
## Summary

Adds deployment workload support to the CRUD benchmarking framework, enabling Telescope to measure K8s Deployment create/verify latency on AKS node pools. Also fixes a race condition in Azure LRO handling for scale operations.

Branch cleanup note: Rebased and squashed from 141 commits → 4 logical commits for reviewability. All prior review feedback has been addressed.

## Changes

**'modules/python/crud/azure/node_pool_crud.py'**

- Add `create_deployment()` — creates N deployments from a YAML template, validates readiness via `wait_for_condition`, and handles partial failures gracefully
- Derive `[label_selector]` from parameters (no hardcoding)
- Use `[self.step_timeout]` instead of hardcoded timeout
- Use `%-style` logging for consistency with existing methods
- Make `namespace` configurable with `"default"` as default value


**'modules/python/crud/main.py'**
- Add `handle_workload_operations()` dispatcher with `deployment` subcommand
- CLI args: `--number-of-deployments`, ``--replicas``, `--manifest-dir`, `[--node-pool-name]`
- Remove unnecessary `[--deployment-name]` argument
- Add `else` clause returning error on unknown workload commands

**'modules/python/crud/workload_templates/deployment.yml'**
- Parameterized deployment manifest with configurable replicas, label_selector, and namespace

`[steps/engine/crud/k8s/execute.yml]`
- Add deployment script block between scale-up and scale-down operations
- Wire `NUMBER_OF_DEPLOYMENTS`, `REPLICAS`, `MANIFEST_DIR` environment variables

 `[steps/topology/k8s-crud-gpu/execute-crud.yml]`
- Pass deployment parameters from pipeline matrix through topology layer

**'modules/python/clients/aks_client.py'**
- Fix: chain `.result()` on `begin_create_or_update()` in `scale_node_pool` and `_progressive_scale` — blocks until ARM completes before proceeding. Matches existing pattern in `create_node_pool` (line 331) and `delete_node_pool` (line 573) which already call `.result()`. Prevents race condition where `wait_for_nodes_ready` runs before ARM has processed the scale request. Consistent with the internal repo which also calls `.result()` on all operations.

## Tests

Unit tests in `test_azure_node_pool_crud.py` and `test_main.py`:

`test_create_deployment_success` — all deployments ready
`test_create_deployment_failure` — all fail to become ready
`test_create_deployment_partial_success` — continues on individual failures, returns False
`test_multiple_deployment`s — verifies N deployments created sequentially
`test_handle_workload_operations` — deployment command routing
`test_handle_workload_operations_unknown_command` — returns error
`test_progressive_scaling_failure` — scale continues after step failure
`test_scale_down_fails_continues` — delete still runs after scale-down error
`test_returns_false_early_exit` — unknown operation returns False

engineeredcurlz committed:
initial update

lokesh-keyan committed:
Merge branch 'main' into test-refactor

lokesh-keyan committed:
wip: add create_deployment function to crud

Commit:
add import for handle_worload_operation function

Commit:
add test for success

Commit:
change operation name

Commit:
update operation name in test

Commit:
add test for failure

Commit:
add exception test

engineeredcurlz committed:
Merge branch 'main' into test-refactor

Commit:
Linting error: removed elif and else

Commit:
Merge branch 'test-refactor' of https://github.com/Azure/telescope into test-refactor

Commit:
fixed the spacing

Commit:
removed extra spaces

Commit:
Add deployment_name for consistency and to reference later

Commit:
verify deployment using wait condition

Commit:
Add logging for maniest and to wait for deployment - debug

Commit:
add logger for deployment success

Commit:
verify pods are available in deployment

Commit:
add failure count

Commit:
add logger to verify deployment

Commit:
add unit test for create_deployment method

Commit:
ran lint

engineeredcurlz committed:
Add test for deployment partial sucess

engineeredcurlz committed:
Add test for multiple deployments

engineeredcurlz committed:
Add test for progressive scaling failure

engineeredcurlz committed:
Add test in node_pool_crud for returns false early exit

engineeredcurlz committed:
Add test in node_pool_crud for scale up fails but continues to scale down

engineeredcurlz committed:
Add test for node_pool_crud for scale down fails operation continues

engineeredcurlz committed:
Add test in node_pool_crud for deployment partial success

engineeredcurlz committed:
pipeline test

engineeredcurlz committed:
linting

engineeredcurlz committed:
yaml lint

engineeredcurlz committed:
add python security dependency

engineeredcurlz committed:
fix dependency

engineeredcurlz committed:
Merge branch 'main' into test-refactor

engineeredcurlz committed:
update

engineeredcurlz committed:
added matrix variables to pipeline

engineeredcurlz committed:
testing: set GPU_NODE_POOL to empty string

engineeredcurlz committed:
update vm size

engineeredcurlz committed:
testing change vm size with available quota

engineeredcurlz committed:
update node count + vm size

engineeredcurlz committed:
update: topology selection

engineeredcurlz committed:
add deployment step after scale-up operation

engineeredcurlz committed:
wire deployment parameters through k8s-crud-gpu topology

engineeredcurlz committed:
correct deployment command routing and kwargs in handle_workload_operations

engineeredcurlz committed:
correct topology name and add deployment matrix variables

engineeredcurlz committed:
update handle_workload_operations tests to match deployment command

engineeredcurlz committed:
fix yamllint and pylint warnings

engineeredcurlz committed:
add correct indentation

engineeredcurlz committed:
iterate multi-doc YAML generator when applying deployment manifests

engineeredcurlz committed:
refactor: seperate deploy workloads into its own pipelinee step

engineeredcurlz committed:
fix: execute k8s workload operations displayname

engineeredcurlz committed:
fix: prevent infinite loop in azure node pool deployment tests

yaml.safe_load_all() enters an infinite loop when passed a MagicMock
object because PyYAML detects the .read attribute and treats it as a
file-like stream, then loops forever waiting to buffer enough bytes
(len(MagicMock()) returns 0 by default).

Fix by setting create_template.return_value to a valid YAML string in
the three create_deployment tests, so yaml.safe_load_all receives a
real string and parses it via the non-blocking code path.

Affected tests:
- test_create_deployment_success
- test_create_deployment_failure
- test_create_deployment_partial_success

engineeredcurlz committed:
Merge branch 'main' into test-refactor

engineeredcurlz committed:
fix: await Azure LRO poller to prevent scale race condition

begin_create_or_update() returns an LROPoller that was being discarded, allowing execution to continue while Azure still had an operation in-progress. Subsequent scale/delete calls were then rejected with OperationNotAllowed.

Fix by calling poller.result() in scale_node_pool and _progressive_scale to block until Azure fully completes each operation before proceeding.

engineeredcurlz committed:
fix: replace hardcoded timeout with self.step_timeout in create_deployment

engineeredcurlz committed:
refactor: convert f-string logger calls to %-style in create_deployment

engineeredcurlz committed:
feat: make label_selector derive from parameter

nginx-container was hardcoded in deployment template and in create deployment method

- add label_selector to parameters
- replace nginx-container in deployment.yaml (label_alue)
- derive label_value from selector
- pass label_selector directly

engineeredcurlz committed:
feat: remove hardcoding add namespace parameter

engineeredcurlz committed:
fix: remove --deployment-name CLI

engineeredcurlz committed:
fix: use hyphen for --number-of-deployments

diamondpowell commented on modules/python/crud/main.py:
```diff
@@ -320,6 +344,32 @@ def main():
+        "deployment", parents=[common_parser], help="create deployments"
+    )
+    deployment_parser.add_argument("--node-pool-name", required=True, help="Node pool name")
+    deployment_parser.add_argument("--deployment-name", required=True, help="Deployment name")
```
its not mentioned in deploy_kwargs. I'll remove it

engineeredcurlz committed:
fix: return error on unknown workload command

engineeredcurlz committed:
revert: restore original docstring line wrapping

engineeredcurlz commented on modules/python/crud/main.py:
```diff
@@ -146,6 +146,30 @@ def handle_node_pool_operation(node_pool_crud, args):
+                "replicas": args.replicas,
+                "manifest_dir": args.manifest_dir,
+                "number_of_deployments": args.number_of_deployments
+            }
```
have added this to stop false successes and return logged errors

engineeredcurlz committed:
Merge branch 'main' into test-refactor

engineeredcurlz committed:
Merge branch 'main' into test-refactor

lokesh-keyan commented on steps/engine/crud/k8s/execute.yml:
```diff
@@ -54,6 +59,35 @@ steps:
+      --result-dir "$RESULT_DIR" \
+      --node-pool-name "$POOL_NAME" \
+      --deployment-name "$DEPLOYMENT_NAME" \
+      --number_of_deployments "$NUMBER_OF_DEPLOYMENTS" \
```
number_of_deployments to number-of-deployments

diamondpowell replied:
Updated to use number-of-deployments and match argparse

engineeredcurlz committed:
revert: testing complete, revert back to original pipeline

engineeredcurlz committed:
fix: use dashes to match argparse argument in main

engineeredcurlz committed:
fix: remove --deployment-name cli (main.p + execute.yml)

engineeredcurlz committed:
Merge branch 'main' into test-refactor

diamondpowell committed:
feat: add deployment workload to CRUD module

Add create_deployment() to NodePoolCRUD that deploys K8s workloads onto
node pools after provisioning. Implements multi-doc YAML manifest parsing,
configurable replica count, and per-deployment readiness validation via
wait_for_condition.

- Add handle_workload_operations() dispatcher in main.py with 'deployment'
  subcommand supporting --number-of-deployments, --replicas, --manifest-dir
- Add deployment.yml workload template with configurable replicas and
  node affinity via label_selector
- Derive label_selector from node pool name parameter (removes hardcoding)
- Return error on unknown workload command

diamondpowell committed:
feat: wire deployment through pipeline topology

Add deployment execution step between scale-up and scale-down in the
k8s CRUD engine pipeline. Parameters (number_of_deployments, replicas,
manifest_dir) flow from pipeline matrix → topology → engine step → main.py.

- Add deployment script block to steps/engine/crud/k8s/execute.yml
- Pass deployment parameters through topology execute-crud.yml
- Deployment runs after scale-up, before scale-down + delete

diamondpowell committed:
fix: await Azure LRO poller to prevent scale race condition

begin_create_or_update() returns an LROPoller that was being discarded
in scale_node_pool and _progressive_scale, allowing execution to continue
while Azure still had an operation in-progress. Subsequent scale/delete
calls were rejected with OperationNotAllowed.

Call poller.result() to block until Azure fully completes each operation
before proceeding. Aligns scale behavior with create_node_pool and
delete_node_pool which already awaited the poller.

diamondpowell committed:
test: add unit tests for deployment CRUD operations

Add comprehensive test coverage for create_deployment and
handle_workload_operations:

- test_create_deployment_success: single deployment with readiness check
- test_create_deployment_partial_success: some deployments fail, operation
  continues and reports partial success
- test_create_deployment_failure: all deployments fail
- test_multiple_deployments: verifies N deployments created sequentially
- test_handle_workload_operations: deployment command routing + kwargs
- test_handle_workload_operations_unknown_command: returns error
- test_progressive_scaling_failure: scale continues after step failure
- test_scale_down_fails_continues: delete still runs after scale-down error
- test_returns_false_early_exit: unknown operation returns False

diamondpowell committed:
refactor: chain .result() directly to match create/delete pattern

diamondpowell committed:
fix: address Copilot review feedback on deployment workload

- Use per-deployment unique labels to avoid selector collision in wait_for_pods_ready
- Resolve template path via __file__ to work regardless of workingDirectory
- Pass namespace to apply_manifest_from_file for consistency with wait calls
- Remove empty spec.template.metadata.name from deployment template
- Fix docstring to match actual method signature
- Remove unused deployment_name parameter from pipeline topology
- Make --manifest-dir conditional (only pass when non-empty)
- Rename test to match actual behavior (deployment, not pod)

diamondpowell committed:
fix: gate deployment step on azure cloud and use non-GPU VM for test pipeline

- Add conditional on deployment step: only runs when cloud=='azure'
- Add hasattr check in handle_workload_operations for unsupported providers
- Switch test pipeline to Standard_D4s_v3 (quota available)

diamondpowell committed:
fix: make --manifest-dir optional (defaults to built-in template)

xinWeiWei24 commented on modules/python/crud/main.py:
```diff
@@ -320,6 +351,32 @@ def main():
+    deployment_parser = subparsers.add_parser(
+        "deployment", parents=[common_parser], help="create deployments"
+    )
+    deployment_parser.add_argument("--node-pool-name", required=True, help="Node pool name")
```
Since the handling function is named handle_workload_operations and is designed to support other operation types in the future, would it be better to name this workload_parser instead of deployment_parser?

diamondpowell replied:
Good suggestion, the docstring already references statefulset and jobs because those are coming in follow-up PRs ([#1132](https://github.com/Azure/telescope/pull/1132) and [#1133](https://github.com/Azure/telescope/pull/1133)). Each workload type has its own parser with unique args. They all route to handle_workload_operations.  Renaming to workload_parser would be ambiguous since there are multiple.

xinWeiWei24 replied:
Looking at the three parsers side-by-side, almost every argument is duplicated:
--node-pool-name, --manifest-dir, --replicas/--completions (same semantics), and the only real difference is the cosmetic name of --number-of-{deployments,statefulsets,jobs}.

I'd still keep three subparsers (so --help stays clean and future workload-specific flags have a natural home), but suggest extracting a workload_common_parser with --node-pool-name, --manifest-dir, --replicas, --count, --label-selector, and reusing it via parents=[common_parser, workload_common_parser]. Each subparser would then only declare what is genuinely workload-specific.

This addresses the duplication without losing the per-workload --help clarity that motivated keeping them separate.

```python
workload_common_parser = argparse.ArgumentParser(add_help=False)
workload_common_parser.add_argument(
    "--node-pool-name", required=True, help="Node pool name"
)
workload_common_parser.add_argument(
    "--manifest-dir", required=True,
    help="Directory containing Kubernetes manifest files for the workload"
)
workload_common_parser.add_argument(
    "--replicas", type=int, default=10,
    help="Number of replicas per workload (for jobs, maps to spec.completions)"
)
workload_common_parser.add_argument(
    "--count", type=int, default=1,
    help="Number of workloads to create"
)
workload_common_parser.add_argument(
    "--label-selector", default="app=nginx-container",
    help="Label selector for created pods (default: app=nginx-container)"
)

# Deployment subcommand
deployment_parser = subparsers.add_parser(
    "deployment",
    parents=[common_parser, workload_common_parser],
    help="create deployments",
)
```

diamondpowell replied:
Ahhhh yeah I see the duplication in each one. Extracted a workload_common_parser with those plus a generic --count for number-of-workloads. Now the deployment parser is just:
```
deployment_parser = subparsers.add_parser(
    "deployment", parents=[common_parser, workload_common_parser], help="create deployments"
)
```
There are now zero unique args. Statefulset and jobs PRs will be the same, just parents=[common_parser, workload_common_parser] and they're done.

Also renamed --number-of-deployments to --count (the other workloads will follow the same) and updated the pipeline YAML chain to match. Made --manifest-dir optional everywhere since all three workloads have built-in templates under workload_templates/.

xinWeiWei24 commented on steps/topology/k8s-crud-gpu/execute-crud.yml:
```diff
@@ -22,3 +22,6 @@ steps:
     result_dir: $(System.DefaultWorkingDirectory)/$(RUN_ID)
     gpu_node_pool: $(GPU_NODE_POOL)
     step_wait_time: $(STEP_WAIT_TIME)
+    number_of_deployments: $(NUMBER_OF_DEPLOYMENTS)
```
Do you need to update it with your custom values in pipelines/perf-eval/GPU Benchmark/k8s-gpu-cluster-crud.yml? I didn’t see any changes here.

diamondpowell replied:
Oh you're right, I didn't set the GPU pipeline matrix setting, which means there were no defaults passing. Added to the azure blocks, skipped aws since deployment is gated to azure

xinWeiWei24 commented on modules/python/crud/azure/node_pool_crud.py:
```diff
@@ -270,3 +272,127 @@ def all(
+                logger.warning("Created %d/%d deployment(s)", successful_deployments, number_of_deployments)
+                return False
+            logger.error("Failed to create any deployments")
+            return False
```
This can be merged like created 0/number_of_deployments

diamondpowell replied:
Good call, merged them. `Created 0/5` gives more info than "Failed to create any" anyway.

xinWeiWei24 commented on modules/python/crud/azure/node_pool_crud.py:
```diff
@@ -270,3 +272,127 @@ def all(
+
+        except Exception as e:
+            logger.error("Failed to create deployments: %s", e)
+            return False
```
Recommend removing the outer try/except entirely and letting unexpected errors propagate. If we want to keep "one bad deployment doesn't stop the rest" behavior, the inner except Exception is enough

diamondpowell replied:
Oh yeah good point, the inner except already handles the per-deployment failures so the outer one was just hiding things. Removed it so any setup issues will actually surface instead of silently returning False.

xinWeiWei24 commented on modules/python/crud/azure/node_pool_crud.py:
```diff
@@ -270,3 +272,127 @@ def all(
+                logger.info("Creating deployment %d/%d", deployment_index, number_of_deployments)
+
+                try:
+                    if manifest_dir:
```
better to extract the per-deployment logic into a helper method

diamondpowell replied:
Yes this does make more sense. Pulled the actual work (resolve template, build name, apply, wait) into _apply_deployment so now the loop just calls it, counts successes, and catches errors. Also made it raise a TimeoutError when the deployment isn't ready instead of returning False,nthat way the loop's except handles everything in one place.

diamondpowell committed:
fix: add deployment defaults to GPU pipeline matrix for Azure stages

diamondpowell committed:
fix: simplify deployment error logging by merging duplicate branches

diamondpowell committed:
refactor: remove redundant outer try/except in create_deployment

diamondpowell committed:
refactor: extract _apply_deployment helper from create_deployment loop

diamondpowell committed:
ci: configure new-pipeline-test.yml for CRUD deployment testing

diamondpowell committed:
fix: only set gpu_profile when gpu_node_pool is true

diamondpowell committed:
refactor: extract workload_common_parser for shared workload args

diamondpowell committed:
revert: restore new-pipeline-test.yml after successful CRUD pipeline test

xinWeiWei24 commented on modules/python/crud/azure/node_pool_crud.py:
```diff
@@ -270,3 +272,138 @@ def all(
             logger.error(error_msg)
             errors.append(error_msg)
             return False
+    def _apply_deployment(
```
Please add a blank line before the function.

diamondpowell replied:
thanks for this catch! added the blank line

xinWeiWei24 commented on modules/python/crud/azure/node_pool_crud.py:
```diff
@@ -270,3 +272,138 @@ def all(
+        )
+
+        logger.info("Successfully created and verified deployment %d", deployment_index)
+    def create_deployment(
```
Please add a blank line before the function.

diamondpowell replied:
added the blank line here too

diamondpowell committed:
style: add blank lines before _apply_deployment and create_deployment methods
